### ISSUE TITLE
torch_dtype=torch.float16  옵션은 잘 동작하고 있습니까? 모델 전처리를 통하여 메모리 요구량을 낮출 수 있습니다. 

### ISSUE THREAD

**Issue by noah-ko** [kakaobrain/kogpt#6]
안녕하세요

우선 모델의 공개에 대하여 진심으로 감사 드립니다. 세부 내용중에 보면 다음과 같은 내용이 있습니다.

GPU
The following is the recommended minimum GPU hardware guidance for a handful of example KoGPT.

half-precision requires NVIDIA GPUS based on Volta, Turing or Ampere
32GB GPU RAM in the required minimum memory size

하지만 모델의 파라메터 사이즈와 자료형을 생각할때 물리적인 용량은 대략 12Gb 근처로 생각됩니다.
그리고 실제 공개하여주신 모델 파일의 tensor 타입은 볼떄 fp32로 저장되어 있는것으로 보입니다.

실제 gpu의 로딩 시점에 볼떄  torch_dtype=torch.float16 구성과 관계없이 fp32로 로딩이 이루어지고 있고 
그로인하여 많은 gpu 메모리를 할당합니다.

아래와 같은 간단한 스크립트를 통하여 모델의 메모리 요구량을 절반 수준으로 줄일 수 있습니다.
제가 언급한 이슈가 연산에 영향이 있었다면 속도도 대폭 개선될 것 입니다.

```
import torch
from transformers import AutoTokenizer, AutoModelForCausalLM

path = "/home/noah/.cache/huggingface/transformers/1386e39caf0b158682709eb063f0231e03f868a0f87846c1eb777a79f161f87d.ce4d05ebacaac5ad33896c20e5373d786588147616bced327805834cb4beaf8f"

model = torch.load(path)

for i in model:
    t = model.get(i).dtype
    if (t == torch.float32):
        print (t)
        model[i] = model.get(i).half()
        
torch.save(model, './fp16.pth')
```

<img width="588" alt="스크린샷 2021-11-19 오후 12 47 35" src="https://user-images.githubusercontent.com/69115576/142562529-99dc31ee-1615-438a-882b-59e92c6b0c0b.png">

**Comment by JoonHong-Kim** [comment]
비슷하게 Trainer Arguments에 fp16 parameter를 True 로 주는 방법도 있습니다.
그리고 단순 fp16만을 사용 시 정밀도가 현저히 떨어지기 때문에 (gradient 값들이 매우 작기 때문)
이 방식을 사용시 automatic mixed precision 방식을 적용하기 때문에 단점을 많이 커버한다 생각합니다.

**Comment by seekseek** [comment]
@JoonHong-Kim 현재 이슈는 학습과는 연관이 없고 amp나 gradient scalar 등 어떤 기법등을 활용하였는지 관련이 없으며 공개기전 확인이 불가능한 영역입니다.

다시 정리하면 현재 이슈는 pytorch 혹은 transformers 의 torch_dtype=torch.float16 옵션의 구현이 임시 메모리 사용량 관점에서 오버헤드가 있고 그로 인하여 kogpt 모델의 gpu 메모리 요구사항이 과다하게 표시되었다는 것입니다.

low_cpu_mem_usage 끄는는것으로 gpu 메모리의 임시 할당 정도를 줄일 수 있지만 이럴 경우 cpu 메모리가 적어도 40Gb 이상 필요한것으로 보입니다.

제가 제안한 방법은 cpu와 gpu 메모리 사용을을 모두 최소화하여 (12Gb 내외) 구동이 가능하기 때문에 배포 파일의 형식의 변경을 제안하는 것 입니다.

**Comment by noah-ko** [comment]
@JoonHong-Kim 현재 이슈는 학습과는 연관이 없고 amp나 gradient scalar 등 어떤 기법등을 활용하였는지 관련이 없으며 공개되기전엔 확인이 불가능한 영역입니다.

다시 정리하면 현재 이슈는 pytorch 혹은 transformers 의 torch_dtype=torch.float16 옵션의 구현이 임시 메모리 사용량 관점에서 오버헤드가 있고 그로 인하여 kogpt 모델의 gpu 메모리 요구사항이 과다하게 표시되었다는 것입니다.

low_cpu_mem_usage 끄는것으로 gpu 메모리의 임시 할당 정도를 줄일 수 있지만 이럴 경우 cpu 메모리가 적어도 40Gb 이상 필요한것으로 보입니다.

제가 제안한 방법은 cpu와 gpu 메모리 사용을을 모두 최소화하여 (12Gb 내외) 구동이 가능하기 때문에 배포 파일의 형식의 변경을 제안하는 것 입니다.
O

**Comment by wbaek** [comment]
@noah-ko 저희가 배포시에 `low_cpu_mem_usage=False`를 통해 GPU메모리를 줄일 수 있는 방법을 누락하여 torch.float16이 정상동작하지 않아 float16을 지원하지 않아 GPU메모리 사용량이 많았습니다.

하지만 제안주신 바가 huggingface GPT-J모델에서 제안된 사용방법에 적합하였고
https://huggingface.co/transformers/master/model_doc/gptj.html#overview

실제로 CPU메모리 사용량 뿐만아니라 모델 다운로드 용량 및 속도까지 전체적으로 모델이 시작하는 시간을 줄일 수 있을것으로 판단되여 float16모델을 신규로 배포하였습니다.
https://huggingface.co/kakaobrain/kogpt/tree/KoGPT6B-ryan1.5b-float16

관련내용은 kogpt inference코드 및 예제 코드에 반영하도록 하겠습니다.
